### PR TITLE
[seek] adds smart seek support for music

### PR DESCRIPTION
**Pull request by xhaggi**
This PR does the following:
- refactor seek handling out of CGUIWindowFullScreen + CApplication and put it in CSeekHandler (done with the action listener pattern)
- drop the AS to configure small step back and forward it to our new smart seeking
- unify seek/skip keys for FullscreenVideo and Visualisation in most of our keymaps
- extends the delay setting to be turned off and drop the validation if only one seek step is enabled to turn off the delay
- adds the seek settings (delay, steps) also to music and differ between in SeekHandler depending on the content (video/audio)

@da-anda @mkortstiege it would be nice if you might take a look.

### CODE REVIEW

**Comment by jjd-uk** [comment]
I've just triggered a multiplatform test build in jenkins with upload to mirrors so I can take look:

http://jenkins.kodi.tv/view/Helpers/job/BuildMulti-PR-Manually/72/

**Comment by xhaggi** [comment]
excellent ;)

File: xbmc/utils/SeekHandler.cpp [hunk 1]
```diff
@@ -208,6 +226,64 @@ void CSeekHandler::OnSettingChanged(const CSetting *setting)
...
+        g_infoManager.SetDisplayAfterSeek();
+      return true;
+    }
+    case ACTION_ANALOG_SEEK_FORWARD:
+    case ACTION_ANALOG_SEEK_BACK:
+    {
+      if (action.GetAmount())
+        Seek(action.GetID() == ACTION_ANALOG_SEEK_FORWARD, action.GetAmount(), action.GetRepeat(), true);
+      return true;
+    }
```
Comment (mkortstiege) [inline xbmc/utils/SeekHandler.cpp]: default switch statement missing

File: xbmc/utils/SeekHandler.cpp [hunk 2]
```diff
@@ -53,46 +58,66 @@ void CSeekHandler::Reset()
...
+  seekTypeSettingMap.insert(std::pair<SeekType, std::string>(SEEK_TYPE_VIDEO, "videoplayer.seeksteps"));
+  seekTypeSettingMap.insert(std::pair<SeekType, std::string>(SEEK_TYPE_MUSIC, "musicplayer.seeksteps"));
+
+  for (std::map<SeekType, std::string>::iterator it = seekTypeSettingMap.begin(); it!=seekTypeSettingMap.end(); ++it)
+  {
+    std::vector<int> forwardSeekSteps;
+    std::vector<int> backwardSeekSteps;
+
+    std::vector<CVariant> seekSteps = CSettings::Get().GetList(it->second);
+    for (std::vector<CVariant>::iterator it = seekSteps.begin(); it != seekSteps.end(); ++it) {
```
Comment (mkortstiege) [inline xbmc/utils/SeekHandler.cpp]: please move the bracket to next line

**Comment by xhaggi** [comment]
@mkortstiege done in fixup commit

**Comment by xhaggi** [comment]
@jjd-uk i've startet a new WIN32 test-build because the multi failed.

**Comment by jjd-uk** [comment]
@xhaggi thanks for kicking off another WIN32 build.

I've had a chance to look at this now and it it all tests fine for me with both videos & music, I really like the changes made here particularly in the music visualisation keymap, so a big thumbs up from me.

**Comment by xhaggi** [comment]
jenkins build this please

**Comment by xhaggi** [comment]
jenkins build this please

**Comment by xhaggi** [comment]
@da-anda any objections, if not i'll push the button

**Comment by da-anda** [comment]
Where you able to fix the issue with changes not getting applied while playing music? (haven't tested the latest version of this PR). If so, it's good to go from my end. 
The only thing that could be changed it the default step sizes for music to "-60,-30,-15,-7,7,15,30,60". +/- 3min is a huge jump for regular songs, so I think it's better if we use +/-1min as max step size and sum it up. Also +7 might make sense here.

Personally I also think the default delay could be 750ms, which should stil be enough for the average user (only if you agree).

**Comment by jjd-uk** [comment]
@da-anda what issue was there with music? I tested with latest version of this and all seemed ok.

**Comment by xhaggi** [comment]
@da-anda I've fixed the callback for music settings yesterday after you reported it. 

>The only thing that could be changed is the default step sizes for music to "-60,-30,-15,-7,7,15,30,60". +/- 3min is a huge jump for regular songs, so I think it's better if we use +/-1min as max step size and sum it up. Also +7 might make sense here.

done

>Personally I also think the default delay could be 750ms, which should stil be enough for the average user (only if you agree).

done

**Comment by da-anda** [comment]
@xhaggi great, thanks

@jjd-uk I was playing music while I changed the step sizes for music seeking and these didn't get applied right away while same worked for video.

**Comment by jjd-uk** [comment]
That explains the difference then, I changed the enabled step sizes before playing anything and all was ok.

**Comment by xhaggi** [comment]
jenkins build and merge

**Comment by da-anda** [comment]
just added the wiki label because the `<seeksteps>` setting needs to be added to the as.xml documentation and the removed as.xml settings need to be marked as "up to version 15 / Helix"

**Comment by NedScott** [comment]
I think I've added it right, but feel free to poke me if any of this needs to be changed: http://kodi.wiki/index.php?title=Advancedsettings.xml&diff=86674&oldid=86661

**Pull request merged**